File: qqbot-protocol/src/main/java/com/mieai/qqbot/protocol/json/JacksonJsonCodec.java
package com.mieai.qqbot.protocol.json;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.json.JsonMapper;
import com.mieai.qqbot.protocol.gateway.GatewayEnvelope;
import java.util.Objects;

final class JacksonJsonCodec implements JsonCodec {
    private final ObjectMapper objectMapper;

    JacksonJsonCodec() {
        objectMapper = JsonMapper.builder()
                .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
                .build();
        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
    }

    @Override
    public String encode(Object value) {
        try {
            return objectMapper.writeValueAsString(value);
        } catch (JsonProcessingException exception) {
            throw new JsonCodecException("Unable to encode protocol JSON", exception);
        }
    }

    @Override
    public <T> T decode(String json, Class<T> type) {
        Objects.requireNonNull(json, "json");
        Objects.requireNonNull(type, "type");
        try {
            return objectMapper.readValue(json, type);
        } catch (JsonProcessingException exception) {
            throw new JsonCodecException("Unable to decode protocol JSON as " + type.getName(), exception);
        }
    }

    @Override
    public <T> GatewayEnvelope<T> decodeGatewayEnvelope(String json, Class<T> dataType) {
        Objects.requireNonNull(json, "json");
        Objects.requireNonNull(dataType, "dataType");
        JavaType envelopeType = objectMapper.getTypeFactory()
                .constructParametricType(GatewayEnvelope.class, dataType);
        try {
            return objectMapper.readValue(json, envelopeType);
        } catch (JsonProcessingException exception) {
            throw new JsonCodecException(
                    "Unable to decode Gateway envelope with data type " + dataType.getName(), exception);
        }
    }
}

File: qqbot-protocol/src/main/kotlin/com/mieai/qqbot/protocol/json/JacksonJsonCodec.kt
package com.mieai.qqbot.protocol.json

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.json.JsonMapper
import com.fasterxml.jackson.module.kotlin.KotlinModule
import com.mieai.qqbot.protocol.gateway.GatewayEnvelope

internal class JacksonJsonCodec : JsonCodec {
    private val objectMapper: ObjectMapper = JsonMapper.builder()
        .addModule(KotlinModule.Builder().build())
        .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
        .build()
        .apply { setSerializationInclusion(JsonInclude.Include.NON_NULL) }

    override fun encode(value: Any?): String = try {
        objectMapper.writeValueAsString(value)
    } catch (exception: JsonProcessingException) {
        throw JsonCodecException("Unable to encode protocol JSON", exception)
    }

    override fun <T> decode(json: String, type: Class<T>): T {
        return try {
            objectMapper.readValue(json, type)
        } catch (exception: JsonProcessingException) {
            throw JsonCodecException("Unable to decode protocol JSON as ${type.name}", exception)
        }
    }

    override fun <T> decodeGatewayEnvelope(json: String, dataType: Class<T>): GatewayEnvelope<T> {
        val envelopeType = objectMapper.typeFactory.constructParametricType(GatewayEnvelope::class.java, dataType)
        return try {
            objectMapper.readValue(json, envelopeType)
        } catch (exception: JsonProcessingException) {
            throw JsonCodecException("Unable to decode Gateway envelope with data type ${dataType.name}", exception)
        }
    }
}
